### ISSUE TITLE
load never fires when setting setRequestInterceptionEnabled to true (and disable JS)

### ISSUE THREAD

**Issue by pocketjoso** [puppeteer/puppeteer#662]
https://github.com/GoogleChrome/puppeteer/pull/565 fixed most cases I found with pages never reporting the load event in puppeteer, but I have here found another url that breaks in the same way - only after calling `setRequestInterceptionEnabled(true)`.

## Environment

Puppeteer: 0.11.0
Node: 7.10.1.
OS: macOS Sierra

### Js disabled + request interception => never loads
This is the real problem that relates to this issue. Originally when I posted I missed the critical detail that JS needs to be disabled for the load to never happen:
```js
const puppeteer = require('puppeteer');
(async() => {
const browser = await puppeteer.launch({
      ignoreHTTPSErrors: true,
      args: ['--disable-setuid-sandbox', '--no-sandbox']
    })
const page = await browser.newPage()

await page.setJavaScriptEnabled(false)
await page.setRequestInterceptionEnabled(true)
page.on('request', intercepted => intercepted.continue())

await page.goto('https://apartmentsdoralfl.com/')
console.log('page load DONE') // never happens
browser.close();
})();
```

**NOTE that the url I'm using in this test case it not the one I had when I originally raised the issue - the original url now works (and along with it about 80% of the problematic url's I had) - but the one listed here does not.

---
Thanks for all the hard work on puppeteer, it's shaping up for me to be able to launch a [new major version of Penthouse](https://github.com/pocketjoso/penthouse/pull/174) with it replacing phantomjs. 👍

**Comment by CommanderXL** [comment]
I have met the same scenes。Has this bug solved?

**Comment by ghost** [comment]
Do you have a page.on request that does request.continue on it? I noticed nothing will load after I enable interception unless I have something similar to
```
page.on('request', req => { req.continue() });
```

**Comment by aslushnikov** [comment]
@pocketjoso indeed, you don't continue requests. The snippet started working for me once i added request continuation.

Speculatively closing this; feel free to write here if it still doesn't work for you.

**Comment by pocketjoso** [comment]
@aslushnikov 

This is still a problem for me. Here's an update:
(also updated first post)

## With intercepted request handling, more are intercepted after page.goto resolves
So I added a request listener to the test case: 
```js
const puppeteer = require('puppeteer');
(async() => {
const browser = await puppeteer.launch({
      ignoreHTTPSErrors: true,
      args: ['--disable-setuid-sandbox', '--no-sandbox']
    })
const page = await browser.newPage()

await page.setRequestInterceptionEnabled(true)
page.on('request', intercepted => intercepted.continue())

await page.goto('https://www.studentdebtrelief.us/')
console.log('page load DONE')
browser.close();
})();
```

Now the page loads, but request(s) come in after the browser.close call has been made, causing `Network.continueInterceptedRequest` Error:
```
page load DONE
(node:32015) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: Protocol error (Network.continueInterceptedRequest): Target closed.
```

I'm not sure whether this is a regular occurance when you run with JS enabled (I normally don't), but it seems to me that this is just the problem of #695 - not really related to this issue here.

However what I describe in the next post is:

**Comment by pocketjoso** [comment]
adding `await page.setJavaScriptEnabled(false)` to the reduced test case, the page never loads. This is probably what was missing in my original reduced test case. In full:

## With js disabled, never loads
```js
const puppeteer = require('puppeteer');
(async() => {
const browser = await puppeteer.launch({
      ignoreHTTPSErrors: true,
      args: ['--disable-setuid-sandbox', '--no-sandbox']
    })
const page = await browser.newPage()

await page.setJavaScriptEnabled(false)
await page.setRequestInterceptionEnabled(true)
page.on('request', intercepted => intercepted.continue())

await page.goto('https://www.studentdebtrelief.us/')
console.log('page load DONE') // never happens
browser.close();
})();
```

**Comment by aslushnikov** [comment]
@pocketjoso ok thanks, I can repro now, will take a look.

**Comment by pocketjoso** [comment]
Just checking in - any info on this yet? Don't want to stress you, just check timings for making my [1.0 release for Penthouse](https://github.com/pocketjoso/penthouse/pull/174) which uses Puppeteer; whether I should wait for this or not.

Thanks for all the hard work on this!

**Comment by addyosmani** [comment]
@aslushnikov Thanks for taking a look 👋  If it helps at all with prioritization, @pocketjoso's work to use Puppeteer in Penthouse is important for the ecosystem as tools like [critical](https://github.com/addyosmani/critical/issues/229) are relying on it for a shift over to Chrome headless.

**Comment by pocketjoso** [comment]
Update from me:

I am still seeing the same problem, but it seems like `0.11` fixed it for _some_ urls.

✅ https://www.studentdebtrelief.us/ - now works
❌ https://apartmentsdoralfl.com/ - _not_ working (never tested before 0.11).

Will update the first post so that it mentions this still non working url.

**Comment by aslushnikov** [comment]
Thanks @pocketjoso for keeping this up-to-date.

The https://apartmentsdoralfl.com/ loads for me on chrome 60 in 2.8 minutes. I started waiting for `networkidle` instead of the load event to complete navigation, and the following works for me:

```js
const puppeteer = require('puppeteer');
(async() => {
  const browser = await puppeteer.launch({
        ignoreHTTPSErrors: true,
        headless: false,
        args: ['--disable-setuid-sandbox', '--no-sandbox']
      })
  const page = await browser.newPage()

  await page.setJavaScriptEnabled(false)
  await page.setRequestInterceptionEnabled(true)
  page.on('request', intercepted => intercepted.continue())

  await page.goto('https://apartmentsdoralfl.com/ ', {timeout: 1000000, waitUntil: 'networkidle'})
  console.log('page load DONE');
  browser.close();
})();

```

**Comment by pocketjoso** [comment]
@aslushnikov Thanks for the update. I see that this works, but why is the behavior so different (i.e. works "as excpected", without changing load trigger) when I don't use setRequestInterceptionEnabled?

```js
const puppeteer = require('puppeteer');
(async() => {
const browser = await puppeteer.launch({
      ignoreHTTPSErrors: true,
      args: ['--disable-setuid-sandbox', '--no-sandbox']
    })
const page = await browser.newPage()

await page.setJavaScriptEnabled(false)

// With these two lines enabled load doesn't fire.
// With them commented it does.
// await page.setRequestInterceptionEnabled(true)
// page.on('request', interceptedRequest => interceptedRequest.continue())

await page.goto('https://apartmentsdoralfl.com/')

console.log('page load DONE')
browser.close();
})();
```

Granted I had a look at this site's waterfall and yeah... it's messy, I can see how it causes problems. I just wish to not have hard to understand side effects like this when I enable request interception..

**Comment by aslushnikov** [comment]
> I just wish to not have hard to understand side effects like this when I enable request interception..

@pocketjoso your last script doesn't work for me right now - the `page.goto` throws a timeout exception for me. From what I see, it looks like website is unreliable by itself and makes things hard to debug. 

Do I understand correctly that this could be closed now?

**Comment by pocketjoso** [comment]
Hmm my last script works reliably for me, even from different locations (via VPN).

I would ideally keep this open as it remains an issue, but I can understand if it's lower priority for you. If you want, close it for now, and I'll re-open if/when I have a few more sites that we can more reliably reproduce the error with - sounds good?

**Comment by aslushnikov** [comment]
@pocketjoso deal!

**Comment by ghost** [comment]
One thing of possible note is that apartmentsdoralfl appears to attempt to read all mousemove and touch and drag events and send them to the server, as part of some "anti spambot firewall" system written in PHP.  It might be throttling future requests or blocking them outright for anything, possibly..

**Comment by speedcom** [comment]
I think this is still an issue (at least to me).
No matter what site I want to go to its hanging till getting timeout.
I'm testing this against `0.11.0` puppeteer and `v8.6.0` node versions.

```
const puppeteer = require('puppeteer');

(async () => {

  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  await page.setRequestInterceptionEnabled(true);

  page.on('load', (...args) => {
    console.log("loaded...")
  })

  page.on('request', async (req) => {
    console.log(req.url);
  });

  await page.goto('https://example.com', {
    timeout: 100000
  });

  await browser.close();
})();

  });

  await page.goto('http://example.com', {
    timeout: 100000
  });

  await browser.close();
})();

```

Without flag being enabled flow is working.

**Comment by geertbrit** [comment]
@pocketjoso : 

> I just wish to not have hard to understand side effects like this when I enable request interception..

Seeing a related issue: enabling `setRequestInterceptionEnabled` messes with the screenshot output. This one is nicely reproducible and *might* hint at the same underlying problem.
 
Since some days have passed, do you perhaps have any more insight in why setting setRequestInterceptionEnabled and letting all requests continue result in a different output? 

See below for a simple to reproduce testcase: 

```
const puppeteer = require('puppeteer');
(async() => {
  const browser = await puppeteer.launch()
  const page = await browser.newPage()

  // toggling this does not have an effect. The problem is consistent regardless. 
  //await page.setJavaScriptEnabled(false);

  // With these lines enabled, screenshot always misses product-image (+ a lot of other stuff when full_page=true)
  // Seems below lines should just be a no-op though
  // await page.setRequestInterceptionEnabled(true)
  // page.on('request', intercepted =>  intercepted.continue());

  await page.goto('http://coolblue.nl')

  await page.screenshot({ path: __dirname + '/coolblue.png' });
  browser.close();
})();
```

**Comment by pocketjoso** [comment]
@aslushnikov just wanted to give an update:
For my own purposes I ended up just skipping the wait for `page.goto` to resolve, after a certain timeout, with the assumption that in these case it’s just a badly configured site that keeps spamming requests, which seems to hold for my use cases.

Hence I don't have this problem anymore.

However I noted in my debugging that the requests that prevent the `load` event from firing were not for page assets (media, css, js, font) and not even ones reaching the `on.('request')` listener... I was just wondering, should such requests (xhr?) really be considered for whether the page has loaded or not? Would make sense to me to, if nothing else, offer a `waitUntil` setting for ~ `pageAssetsLoaded` that would ignore (other) xhr requests. But like I said, for my use case I can live without it.

**Comment by aslushnikov** [comment]
Thanks @pocketjoso for the update.

>  I was just wondering, should such requests (xhr?) really be considered for whether the page has loaded or not?

It depends on the situation; module loaders, for example, rely heavily on xhr's.

**Comment by worudso** [comment]
I met a similar case.
```javascript
const puppeteer = require("puppeteer");

(async () => {
  const browser = await puppeteer.launch({headless: false});
  const page = await browser.newPage();

  await page.setJavaScriptEnabled(false);


  await page.setRequestInterception(true);
  page.on("request", async (request) => {
    await request.continue();
  });

  page.on("response", async (response) => {
    if ( response.url.startsWith("https://mdskip.taobao.com") ) {
      console.log(await response.text());
    }
  });

  await page.goto("http://world.tmall.com/item/538687695440.htm");

  await browser.close();
})();
```
Without request interception, it prints proper response. But enabling it, it prints empty string, which means response.text() returns nothing.

**Comment by aslushnikov** [comment]
@worudso the script prints nothing for me both with and without request interception; I tried on puppeteer@next. Please file a new issue if you think something doesn't work and there's a good repro.

**Comment by TheGP** [comment]
So in some many years they didnt fix it?
